File: web/src/crypto/signature.ts
import { concat, utf8 } from './bytes';
import type { EntityRef } from './envelope';
import { type Identity, sign, verify } from './identity';

/** Raw ECDSA P-256, r||s. WebCrypto produces exactly this; DER would be the other form. */
export const SIGNATURE_LENGTH = 64;

/**
 * What an author signs.
 *
 * The ciphertext alone is not enough. Signing it bare would let the server move a signed
 * body onto a different note, or replay an old one as the current version, and the
 * signature would still check out. So the digest covers the slot as well: which vault,
 * which note, under which scope and version, at which sequence.
 *
 * The nonce is in there too — without it, two versions with identical plaintext would
 * produce the same digest and one signature would cover both.
 */
export function revisionPayload(ref: EntityRef, contentSeq: number, sealed: {
  ciphertext: Uint8Array;
  nonce: Uint8Array;
}): Uint8Array {
  const header = utf8(
    `shelf/sig/v1|${ref.vaultId}|${ref.entityId}|${ref.scopeClientId}|${ref.keyVersion}|${contentSeq}|`,
  );

  return concat(header, sealed.nonce, sealed.ciphertext);
}

/**
 * Signs a body. It is what makes "written by" a fact rather than a claim: view, comment
 * and edit are one key, so any reader could otherwise produce ciphertext that decrypts and
 * no reader could tell the difference.
 */
export async function signRevision(
  identity: Identity,
  ref: EntityRef,
  contentSeq: number,
  sealed: { ciphertext: Uint8Array; nonce: Uint8Array },
): Promise<Uint8Array> {
  return sign(identity, revisionPayload(ref, contentSeq, sealed));
}

/** How a stored revision's authorship checks out, or fails to. */
export type Authorship = 'valid' | 'invalid' | 'unsigned' | 'unknown-author';

/**
 * Checks who wrote a revision.
 *
 * `unsigned` is not the same as `invalid`: a body written before signatures existed simply
 * carries none, and calling that forged would be a lie. It is not proof of authorship
 * either, and the view says so rather than showing a name.
 */
export async function checkAuthorship(
  authorPublicBlob: Uint8Array | null,
  signature: Uint8Array | null,
  ref: EntityRef,
  contentSeq: number,
  sealed: { ciphertext: Uint8Array; nonce: Uint8Array },
): Promise<Authorship> {
  if (!signature || signature.length !== SIGNATURE_LENGTH) return 'unsigned';
  if (!authorPublicBlob || authorPublicBlob.length === 0) return 'unknown-author';

  try {
    const ok = await verify(authorPublicBlob, signature, revisionPayload(ref, contentSeq, sealed));

    return ok ? 'valid' : 'invalid';
  } catch {
    // A public key that will not import is not a failed check — it is an author this
    // client cannot make sense of, and saying "forged" would be the wrong accusation.
    return 'unknown-author';
  }
}

File: internal/envelope/signature.go
package envelope

import (
	"crypto/ecdsa"
	"crypto/rand"
	"crypto/sha256"
	"fmt"
	"strconv"
)

// RevisionPayload is what the author of a body signs.
//
// The ciphertext alone would not be enough: signing it bare would let a hostile server move
// a signed body onto a different note, or replay an old one as the current version, and the
// signature would still check out. So the digest covers the slot as well — which vault,
// which note, under which scope and version, at which sequence — and the nonce with it,
// because two versions with identical plaintext would otherwise share one signature.
func RevisionPayload(ref EntityRef, contentSeq int64, sealed Sealed) []byte {
	header := "shelf/sig/v1|" +
		strconv.FormatInt(ref.VaultID, 10) + "|" +
		ref.EntityID + "|" +
		ref.ScopeClientID + "|" +
		strconv.FormatInt(int64(ref.KeyVersion), 10) + "|" +
		strconv.FormatInt(contentSeq, 10) + "|"

	payload := make([]byte, 0, len(header)+len(sealed.Nonce)+len(sealed.Ciphertext))
	payload = append(payload, header...)
	payload = append(payload, sealed.Nonce...)

	return append(payload, sealed.Ciphertext...)
}

// SignRevision signs a body the connector is about to write.
//
// It is what makes "written by" a fact rather than a claim: view, comment and edit are one
// key, so any reader could otherwise produce ciphertext that decrypts, and no reader could
// tell the difference. A connector that wrote unsigned bodies would be indistinguishable
// from a server forging them.
//
// contentSeq is the sequence the write will land on, which is one past what the caller
// last saw — the same number the If-Match precondition carries.
func SignRevision(identity *Identity, ref EntityRef, contentSeq int64, sealed Sealed) ([]byte, error) {
	return Sign(identity.Sign, RevisionPayload(ref, contentSeq, sealed))
}

// Sign produces the raw r || s form WebCrypto reads. ecdsa.SignASN1 would produce DER,
// which every existing signature in this system is not.
func Sign(private *ecdsa.PrivateKey, payload []byte) ([]byte, error) {
	digest := sha256.Sum256(payload)

	r, s, err := ecdsa.Sign(rand.Reader, private, digest[:])
	if err != nil {
		return nil, fmt.Errorf("sign: %w", err)
	}

	half := SignatureLength / 2

	signature := make([]byte, SignatureLength)
	r.FillBytes(signature[:half])
	s.FillBytes(signature[half:])

	return signature, nil
}
